Add tutorial on how to supplement rosdep with custom keys (#6187)
ros2/ros2_documentation: source/Tutorials/Advanced/Supplementing-Custom-Rosdep-Keys.rst

Anyone can request for new rosdep keys to be added by `contributing to rosdistro <https://github.com/ros/rosdistro/blob/master/CONTRIBUTING.md#rosdep-rules-contributions>`_.
This is the preferred course of action when you have some dependency (like a ``apt`` or ``pip`` package) that you wish to be able to install via ``rosdep``.

However, there are many cases when contributing your keys directly might be difficult.
For example, if the dependency

1. is not available on the target distributions default APT (or pip) repositories
2. is a proprietary library
3. is some niche library which is useful only for you or your organization
4. is a ROS package you :doc:`built and packaged yourself <../../How-To-Guides/Building-a-Custom-Deb-Package>`, but don't wish to share with the broader ROS community

While the option exists to :doc:`fork rosdistro <../../How-To-Guides/Using-Custom-Rosdistro>` in its entirety, this might be overkill if you just want to keep using a official ROS distribution as usual, only with some extra rosdep keys defined on top.

> This tutorial explains how to achieve this.

Reviewer: ```suggestion
This tutorial explains how to achieve this.
As a word of warning though, please don't use this indiscriminately.
It can cause very hard to debug problems.
And if you're asking anyone for help on a system with this enabled, make sure to explain everything that's being added or overridden.
```

Author: Good point, thanks.

Can we quantify "very hard to debug problems" a bit more? I guess something like

> binary incompatibilities or version shadowing issues, which can be very hard to debug

?

**Edit**: applied in 04bd74984d80dc3e629e97467d47ac1324bb500b, but let me know what you think about the extension!

Reviewer: Good idea, extending it in another comment.

Now all you have to do is add ``<depend>awesome_library</depend>`` to your ROS packages ``package.xml``, and ``rosdep`` will know how to install the dependency!

Closing remarks
---------------

The toy example above only hints at what is possible with custom rosdep keys.

- **Is your dependency a apt package hosted in a third party PPA?**
  Not a problem.
  Since all ``rosdep`` does is converting the key to a ``apt install`` invocation, ``apt`` will have no problem installing the package (provided you have added the PPA).
- **Is your dependency a pip package hosted in a third party index?**
  Add the index to your ``pip.conf`` and you're good to go.
- **Sources files don't have to point to files on the local machine.**
  Both ``file://`` and ``https://`` syntax is supported.
- **Sources are loaded in alphabetical order.**
  If you add a conflicting rule in the 30 prefix it will not be used.
  If you created a sources file with a 10 prefix it will override packages in the default list (prefix 20).
  If you're using anything from binaries it's highly recommended not to override dependency declarations as it will likely cause binary incompatibilities which can be very hard to debug.

> 

Reviewer 2: ```suggestion
- **Merging keys is not possible.**
  It is not possible to e.g. only add a ``fedora`` installation rule to an existing rosdep key. Depending on the load order, such rule would either be ignored, or it would completely override the whole rosdep key, removing all other installers.

```

Anyone can request for new rosdep keys to be added by `contributing to rosdistro <https://github.com/ros/rosdistro/blob/master/CONTRIBUTING.md#rosdep-rules-contributions>`_.
This is the preferred course of action when you have some dependency (like a ``apt`` or ``pip`` package) that you wish to be able to install via ``rosdep``.

However, there are many cases when contributing your keys directly might be difficult.
For example, if the dependency

1. is not available on the target distributions default APT (or pip) repositories
2. is a proprietary library
3. is some niche library which is useful only for you or your organization
4. is a ROS package you :doc:`built and packaged yourself <../../How-To-Guides/Building-a-Custom-Deb-Package>`, but don't wish to share with the broader ROS community

While the option exists to :doc:`fork rosdistro <../../How-To-Guides/Using-Custom-Rosdistro>` in its entirety, this might be overkill if you just want to keep using a official ROS distribution as usual, only with some extra rosdep keys defined on top.
This tutorial explains how to achieve this.

> As a word of warning though, please don't use this indiscriminately.

Reviewer 3: This is a suggestion, not necessarily a blocking requirement, but in conversations with our new documentation contractors they suggested more liberal use of some of the features available in restructured text. We may want to use a warning (e.g. `.. warning::`) here to improve readability.

> It can cause very hard to debug problems.

Reviewer: ```suggestion
It can cause very hard to debug problems because of binary incompatibilities which can manifest in silent failures, unexplained crashes, or data corruption.
```

Preliminaries: How ``rosdep`` fetches rosdep keys
-------------------------------------------------

In order to get a good understanding of what we're about to do, let's first explore some relevant details about how ``rosdep`` works.

``rosdep`` is similar to other tools like ``apt`` that use a sources list to maintain a local index.
These sources are stored in ``/etc/ros/rosdep/sources.list.d``.
This is similar to how apt stores repositories in ``/etc/apt/sources.list.d``.

By default (as part of first-time setup, ``rosdep init``), you only have a single sources file: ``/etc/ros/rosdep/sources.list.d/20-default.list``.
Inspecting its contents, you will see entries like these:

.. code-block:: console

   $ cat /etc/ros/rosdep/sources.list.d/20-default.list
   ...
   yaml https://raw.githubusercontent.com/ros/rosdistro/master/rosdep/base.yaml
   yaml https://raw.githubusercontent.com/ros/rosdistro/master/rosdep/python.yaml
   ...

These entries are what dictates where ``rosdep`` fetches rosdep keys and their mappings (rosdep **rules**) from when ``rosdep update`` is called.

> When called, ``rosdep`` compiles the relevant contents from all declared entries in all sources files into a local cached index.

Reviewer 3: ```suggestion
When called, ``rosdep`` fetches the relevant contents from all declared entries in all sources files into a local cached index.
```

A small clarity improvement.

.. code-block:: yaml

  awesome_library:
    ubuntu: [awesome_library]
  that_other_library:
    ubuntu:
      pip:
        packages: [another_library]

This defines two new rosdep rules:

1. The key ``awesome_library``, defined only for Ubuntu, mapping to the ``apt`` package of the same name
2. The key ``that_other_library``, defined only for Ubuntu, mapping to the ``pip`` package named ``another_library``

After running ``rosdep update``, ``rosdep`` will detect the new ``30-custom.list``, prompting it to scan the contents of the ``custom_rules.yaml`` file.
Now ``rosdep`` is set up to recognize these new keys and what they should map to:

.. code-block:: console

   $ rosdep resolve awesome_library
   #apt
   awesome_library
   $ rosdep resolve that_other_library
   #pip
   another_library

Now all you have to do is add ``<depend>awesome_library</depend>`` to your ROS packages ``package.xml``, and ``rosdep`` will know how to install the dependency!

Closing remarks
---------------

The toy example above only hints at what is possible with custom rosdep keys.

- **Is your dependency an APT package hosted in a third party PPA?**
  Not a problem.

>   Since all ``rosdep`` does is converting the key to a ``apt install`` invocation, APT will have no problem installing the package (provided you have added the PPA).

Reviewer 3: ```suggestion
  Since all ``rosdep`` does is converting the key to a ``apt install`` invocation, APT will have no problem installing the package (provided you `have added the PPA <https://help.ubuntu.com/stable/ubuntu-help/addremove-ppa.html.en>`_ ).
```

Add some bread crumbs for the less familiar. =)

- **Is your dependency a pip package hosted in a third party index?**
  Add the index to your ``pip.conf`` and you're good to go.
- **Sources files don't have to point to files on the local machine.**
  Both ``file://`` and ``https://`` syntax is supported (on Linux, absolute paths start with ``/``, which leads to a triple slash like ``file:///etc/rosdep/my.file``).
- **Sources are loaded in alphabetical order.**
  If you add a conflicting rule in the 30 prefix it will not be used.
  If you created a sources file with a 10 prefix it will override packages in the default list (prefix 20).
  If you're using packages installed from binary repositories it's highly recommended to not override dependency declarations as it will likely cause binary incompatibilities which can be very hard to debug.
- **Merging keys is not possible.**

>   It is not possible to e.g. only add a ``fedora`` installation rule to an existing rosdep key.

Reviewer 3: ```suggestion
  For example, it is not possible to add only a ``fedora`` installation rule to an existing rosdep key.
```

This might be a bit clearer.

  Depending on the load order, such rule would either be ignored, or it would completely override the whole rosdep key, removing all other installers.

Further reading
---------------

> - https://docs.ros.org/en/independent/api/rosdep/html/rosdep_yaml_format.html

Reviewer 3: ```suggestion
- `rosdep YAML format documentation <https://docs.ros.org/en/independent/api/rosdep/html/rosdep_yaml_format.html>`_
```

A little bit cleaner.

> - https://docs.ros.org/en/independent/api/rosdep/html/contributing_rules.html

Reviewer 3: ```suggestion
- `Contributing rosdep rules <https://docs.ros.org/en/independent/api/rosdep/html/contributing_rules.html>`_
```

A little bit cleaner.
